File: daily_challenges/166.cpp
//T.C : O(denominator) approximately
//S.C : O(1)

class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0)
            return "0";
        string result;
        if ((long long)numerator * (long long)denominator < 0)
            result += '-';
        long long absNumerator = labs(numerator);
        long long absDenominator = labs(denominator);

        long long integerDiv = absNumerator / absDenominator;
        result += to_string(integerDiv); //log(10(integerDiv))

        long long remainder = absNumerator % absDenominator;
        if (remainder == 0)
            return result;
        result += '.';

        unordered_map<int, int> mp;
        // 0 <= remainder <= denominator
        // O(denominator + length) [insert only once]
        while (remainder != 0) {
            if (mp.count(remainder)) {
                result.insert(mp[remainder], "("); //O(length)
                result += ")";
                break;
            }
            mp[remainder] = result.length();
            remainder *= 10;
            int digit = remainder / absDenominator;
            result += to_string(digit); //O(1)

            remainder = remainder % absDenominator;
        }
        return result;
    }
};

File: daily_challenge_leetcode/166.js
/**
 * @param {number} numerator
 * @param {number} denominator
 * @return {string}
 */
var fractionToDecimal = function(numerator, denominator) {
    if(numerator === 0){
        return "0"
    }
    const result = []
    if(numerator * denominator < 0){
        result.push("-")
    }
    const absNum = Math.abs(numerator)
    const absDen = Math.abs(denominator)

    const integerDiv = Math.floor(absNum / absDen)
    result.push(integerDiv.toString())

    let remainder = absNum % absDen
    if(remainder === 0){
        return result.join('')
    }
    result.push('.')
    const mp = new Map()
    while(remainder !== 0){
        if(mp.has(remainder)){
            result.splice(mp.get(remainder), 0, '(');
            result.push(')')
            break
        }
        mp.set(remainder, result.length)
        remainder *= 10
        const digit = Math.floor(remainder / absDen)
        result.push(digit.toString())
        remainder %= absDen
    }
    return result.join('')
};